subprojects/dependency-management/src/test/groovy/org/gradle/api/internal/artifacts/transform/TransformationResultWriterTest.groovy

```diff
@@ -0,0 +1,137 @@
+            outputDir
+        )
+    }
+
+    def "reads and writes output only transformation results"() {
+        expect:
+        assertCanWriteAndReadResult(
+            outputDir.file("first"),
+            outputDir.file("second"),
+            outputDir
+        )
+    }
+
+    def "reads and writes input only transformation results"() {
+        expect:
+        assertCanWriteAndReadResult(
+            inputArtifact.file("inside"),
+            inputArtifact,
+            inputArtifact
+        )
+    }
+
+    def "resolves files in input artifact relative to input artifact"() {
+        def writer = new TransformationResultWriter(inputArtifact, outputDir)
+        def newInputArtifact = file("newInputArtifact").createDir()
+
+        ImmutableList<File> result = ImmutableList.of(
+            inputArtifact.file("inside"),
+            inputArtifact,
+            outputDir,
+            inputArtifact
+        )
+        ImmutableList<File> resultResolvedForNewInputArtifact = ImmutableList.of(
+            newInputArtifact.file("inside"),
+            newInputArtifact,
+            outputDir,
+            newInputArtifact
+        )
+
+        when:
+        def initialResults = writer.writeToFile(resultFile, result)
+        then:
+        resultFile.exists()
+        initialResults.resolveOutputsForInputArtifact(inputArtifact) == result
+        initialResults.resolveOutputsForInputArtifact(newInputArtifact) == resultResolvedForNewInputArtifact
+
+        when:
+        def loadedResults = writer.readResultsFile(resultFile)
+        then:
+        loadedResults.resolveOutputsForInputArtifact(inputArtifact) == result
+        loadedResults.resolveOutputsForInputArtifact(newInputArtifact) == resultResolvedForNewInputArtifact
+    }
+
+    def "loads files in output directory relative to output directory"() {
+        def writer = new TransformationResultWriter(inputArtifact, outputDir)
+        def newOutputDir = file("newOutputDir").createDir()
+
+        ImmutableList<File> result = ImmutableList.of(
+            inputArtifact,
+            outputDir,
+            outputDir.file("output.txt")
+        )
+        ImmutableList<File> resultInNewOutputDir = ImmutableList.of(
+            inputArtifact,
+            newOutputDir,
+            newOutputDir.file("output.txt")
+        )
+
+        when:
+        def initialResults = writer.writeToFile(resultFile, result)
+        then:
+        resultFile.exists()
+        initialResults.resolveOutputsForInputArtifact(inputArtifact) == result
+
+        when:
+        def writerWithNewOutputDir = new TransformationResultWriter(inputArtifact, newOutputDir)
+        def loadedResults = writerWithNewOutputDir.readResultsFile(resultFile)
+        then:
+        loadedResults.resolveOutputsForInputArtifact(inputArtifact) == resultInNewOutputDir
+    }
+
+    private void assertCanWriteAndReadResult(File... files) {
+        def writer = new TransformationResultWriter(inputArtifact, outputDir)
```

Comment:
This could be a field instead of created for each test individually.

Reply:
Made it a field.